File: packages/shared/src/schemas/enriched-ad.ts
import { z } from "zod";
import { AdNetwork } from "./common";

/**
 * Enriched Ad Record (§4.3) — the per-ad output of the MAP step.
 *
 * Each scraped competitor ad is enriched: text through the LLM, creatives through
 * the vision model, into this structured record. Many of these are then reduced
 * into a single `analysis_object`.
 *
 * Reliability note (§4.3): text/format/dates are solid; impression ranges are
 * wide buckets (tier signal only). EVERY field nullable — scraped data is
 * frequently incomplete and must degrade gracefully (§10).
 */

export const ScaleTier = z.enum(["low", "medium", "high"]);
export type ScaleTier = z.infer<typeof ScaleTier>;

export const TextDensity = z.enum(["low", "medium", "high"]);
export type TextDensity = z.infer<typeof TextDensity>;

/** Creative / design-language attributes — Display only (§4.3). */
export const CreativeAttributes = z.object({
  faces: z.boolean().nullable(),
  before_after: z.boolean().nullable(),
  product_vs_lifestyle: z.enum(["product", "lifestyle", "mixed"]).nullable(),
  text_density: TextDensity.nullable(),
  tone: z.string().nullable(),
  dominant_colors: z.array(z.string()).nullable(),
});
export type CreativeAttributes = z.infer<typeof CreativeAttributes>;

export const EnrichedAdRecord = z.object({
  ad_id: z.string(),
  advertiser: z.string().nullable(),
  source: z.enum(["google_transparency", "serpapi", "meta", "cached"]).nullable(),

  // raw captured content
  format: z.enum(["search", "display", "video", "unknown"]).nullable(),
  headline: z.string().nullable(),
  body: z.string().nullable(),
  image_url: z.string().nullable(),
  landing_url: z.string().nullable(),

  // Performance-proxy signals (most valuable, all derived) — §4.3
  first_shown: z.string().nullable(),
  last_shown: z.string().nullable(),
  /** last_shown − first_shown; best free proxy for "is it working". */
  days_running: z.number().int().nonnegative().nullable(),
  still_active: z.boolean().nullable(),
  /** conviction/budget signal */
  variant_count: z.number().int().nonnegative().nullable(),
  /** from impression bucket × region spread — tier only, never precise. */
  scale_tier: ScaleTier.nullable(),

  // Offer & pricing
  offer_type: z.string().nullable(),
  price_points: z.array(z.string()).nullable(),
  promotion_cadence: z.string().nullable(),

  // Messaging & angle
  primary_value_prop: z.string().nullable(),
  emotional_hook: z.string().nullable(),
  implied_persona: z.string().nullable(),
  claims: z.array(z.string()).nullable(),
  cta_verb: z.string().nullable(),

  // Keyword & targeting
  repeated_phrases: z.array(z.string()).nullable(),
  regions: z.array(z.string()).nullable(),
  platforms: z.array(AdNetwork).nullable(),

  // Creative / design language (Display)
  creative: CreativeAttributes.nullable(),
});
export type EnrichedAdRecord = z.infer<typeof EnrichedAdRecord>;

File: services/ai/src/gaa_ai/schemas/enriched_ad.py
"""EnrichedAdRecord + RawAd — mirror packages/shared/src/schemas/enriched-ad.ts.

RawAd is what a scraper returns (pre-enrichment). EnrichedAdRecord is the per-ad
output of the MAP step. Every enriched field is Optional — scraped data is
frequently incomplete and must degrade gracefully (§4.3 reliability note).
"""

from __future__ import annotations

from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from gaa_ai.schemas.common import AdNetwork, ScaleTier, TextDensity


class RawAd(BaseModel):
    """Pre-enrichment ad as returned by a scraper / fixture."""

    model_config = ConfigDict(extra="ignore")

    ad_id: str
    advertiser: str | None = None
    source: Literal["google_transparency", "serpapi", "meta", "cached"] | None = None
    format: Literal["search", "display", "video", "unknown"] | None = None
    headline: str | None = None
    body: str | None = None
    image_url: str | None = None
    landing_url: str | None = None
    first_shown: date | None = None
    last_shown: date | None = None
    still_active: bool | None = None
    variant_count: int | None = None
    impressions_bucket: str | None = None
    regions: list[str] | None = None


class CreativeAttributes(BaseModel):
    """Display-only design-language attributes (vision-tagged)."""

    faces: bool | None = None
    before_after: bool | None = None
    product_vs_lifestyle: Literal["product", "lifestyle", "mixed"] | None = None
    text_density: TextDensity | None = None
    tone: str | None = None
    dominant_colors: list[str] | None = None


class EnrichedAdRecord(BaseModel):
    ad_id: str
    advertiser: str | None = None
    source: Literal["google_transparency", "serpapi", "meta", "cached"] | None = None

    format: Literal["search", "display", "video", "unknown"] | None = None
    headline: str | None = None
    body: str | None = None
    image_url: str | None = None
    landing_url: str | None = None

    # Performance-proxy signals (derived).
    first_shown: date | None = None
    last_shown: date | None = None
    days_running: int | None = Field(default=None, ge=0)
    still_active: bool | None = None
    variant_count: int | None = Field(default=None, ge=0)
    scale_tier: ScaleTier | None = None

    # Offer & pricing.
    offer_type: str | None = None
    price_points: list[str] | None = None
    promotion_cadence: str | None = None

    # Messaging & angle.
    primary_value_prop: str | None = None
    emotional_hook: str | None = None
    implied_persona: str | None = None
    claims: list[str] | None = None
    cta_verb: str | None = None

    # Keyword & targeting.
    repeated_phrases: list[str] | None = None
    regions: list[str] | None = None
    platforms: list[AdNetwork] | None = None

    # Creative / design language (Display).
    creative: CreativeAttributes | None = None
